docs/tutorials-redeliver-messages.md

```diff
@@ -0,0 +1,236 @@
+Pulsar provides at-least-once semantics when the client does not enable transaction because the client may cache some
+messages out of Pulsar when redelivering messages.
+
+Pulsar Consumer API provides four ways to reconsume the unacknowledged messages later:
+
+* **ackTimeout**
+* **deadLetterPolicy**
+  * **reconsumeLaterCumulative**
+  * **reconsumeLater**
+* **negativeAcknowledge**
+* **redeliverUnacknowledgedMessages**
+
+The **ackTimeout** is an automatic protection mechanism. If a consumer configured ackTimeout, the messages will be
+auto-redelivered when the received messages are not acknowledged after a long time. It works at the client side, ensuring
+the unacknowledged messages will be redelivered to another consumer when the connection of a consumer remains active
+but the business system gets stuck or misses to ack some messages. If the consumer gets disconnected when the client crashes,
+the messages will be auto-redelivered by the broker too. However, this mechanism does not wait for the response of acknowledgment,
+so if an acknowledgment fails on the broker side or proxy side, an ack hole may occur.
+
+The **deadLetterPolicy** is a policy in the message queue used to handle messages that cannot be processed properly.
+In many message queue systems (such as RabbitMQ, Google Pub/Sub, Apache Pulsar, etc.), this strategy is implemented.
+In Pulsar, the deadLetterPolicy is implemented at the client side, it creates a new retry letter topic and dead letter
+topic when building a consumer with `deadLetterPolicy` configuration. When a consumer calls `reconsumeLaterCumulative`
+or `reconsumeLater`, the message (method parameter) will be produced to the retry letter topic until the retry time reaches
+the `maxRedeliverCount`. The message will be produced to the dead letter topic when the retry time reaches the `
+maxRedeliverCount`. The main difference between them is that `reconsumeLaterCumulative` will cumulative ack the
+message (method parameter) after it is produced and `reconsumeLater` will individual ack the message (method parameter)
+after it is produced.
+
+The `negativeAcknowledge` is used to redeliver certain unacknowledged messages while `redeliverUnacknowledgedMessages`
+is used to redeliver all the unacknowledged messages received by this consumer. The main difference between them and
+deadLetterPolicy is that there is no new topic created, and there is an unlimited number of redeliveries.
+
+## Best Practice Suggestion
+
+Different scenarios require different best practices. Users who value the order of partition messages and wish to batch
+process data should choose or implement an appropriate routerPolicy to send a batch of ordered messages to the same
+partition. They should also select either Exclusive or Failover subscription modes. For users who do not care about
+message order and those in stream processing scenarios, they can opt to use Shared and Key-shared subscription modes.
```

Comment:
do not care about message order *and* those in stream processing scenario (which usually do care about order) - this is a contradiction

Reply:
You're correct, stream processing scenarios indeed often care about the order of messages, and batch processing scenarios also pay attention to the order of messages. However, stream processing scenarios use individual ack to acknowledge individual messages, while batch processing scenarios use cumulative ack to acknowledge messages in bulk. The subscription types Shared and key-shared recommend using individual ack and do not guarantee the order of messages, so we recommend using Shared and key-shared subscription types in stream processing scenarios where the order of messages is not a concern. 

This is not to say that stream processing scenarios do not care about the order of data, but rather that Shared and key-shared types should be used in stream processing scenarios where the order of data is not a concern.